### PR TITLE
Coral 2.0

### PR DESCRIPTION
**Pull request by suhas-r**

### CODE REVIEW

**Review by jluebeck — APPROVED** [review]

File: coral/path_constraints.py
Comment (suhas-r) [reply coral/path_constraints.py:1]: Yes good catch, messy rebase

File: coral/breakpoint/breakpoint_types.py
Comment (suhas-r) [reply coral/breakpoint/breakpoint_types.py:26]: There's always a header row in the inputs we expect here, which is what this previous [conditional](https://github.com/AmpliconSuite/CoRAL/blob/main/src/infer_breakpoint_graph.py#L85) was avoiding. That added an extra level of nesting to the below logic that I believe makes the code less readable, which was why I preferred this approach.

File: coral/breakpoint/graph.py
Comment (suhas-r) [reply coral/breakpoint/graph.py:1]: bad rebase, removing this file

File: coral/breakpoint/infer_breakpoint_graph.py
Comment (suhas-r) [reply coral/breakpoint/infer_breakpoint_graph.py:13]: follows [Python import grouping style](https://peps.python.org/pep-0008/#imports)

File: coral/cnv_seed.py
Comment (suhas-r) [reply coral/cnv_seed.py:20]: Good catch, this was duplicated in the existing codebase:
https://github.com/AmpliconSuite/CoRAL/blob/281e7add9d5d7f39dba7de53b3498198187602b7/src/cnv_seed.py#L11
https://github.com/AmpliconSuite/CoRAL/blob/281e7add9d5d7f39dba7de53b3498198187602b7/src/breakpoint_utilities.py#L11

Refactored this reference to use the fn defined in `breakpoint_utilities`

File: coral/datatypes.py
Comment (suhas-r) [reply coral/datatypes.py:60]: Sure, agree that it's probably misleading

File: coral/models/utils.py
Comment (suhas-r) [reply coral/models/utils.py:1]: updated to `cycle_utils.py`

File: coral/breakpoint/breakpoint_types.py
Comment (jluebeck) [reply coral/breakpoint/breakpoint_types.py:26]: In prior changes I had left open the option that the user provides a bed file that isn't a .cns produced by CNVkit. This is reflected in the code with the `is_cns` variable. Agreed with Matt that in either case we shouldn't assume a header. We should only skip the header if one exists - not dangerously force a skip without checking what kind of file we're reading.

I suggest modifying so that if it is a .cns file, we skip the first line (or more generically the line that starts with "chromosome"), and also we could do something like skip any line that starts with "#", since that's a general header indicator in bfx. Thoughts?

File: coral/breakpoint/breakpoint_types.py
Comment (suhas-r) [reply coral/breakpoint/breakpoint_types.py:20]: Primarily for simplicity, since we can have click/typer pass us the file object from the CLI and do validation on it before getting to "CoRAL logic", instead of having validation embedded into functions like these as it previously was or potentially causing confusing errors while we're mid-BreakpointGraph generation. It also means we're using the same file at all points in the code, as the same ".bam" / ".cns" could be edited by some external program/user over a single CoRAL process's lifetime. https://click.palletsprojects.com/en/stable/arguments/#file-opening-safety

File: coral/constants.py
Comment (suhas-r) [reply coral/constants.py:1]: I think we should revisit this along with the `ReferenceGenome` enum when we have a clearer idea of how we want to support multiple genomes

File: coral/breakpoint/breakpoint_graph.py
Comment (suhas-r) [reply coral/breakpoint/breakpoint_graph.py:1]: Moved all auxiliary BP functions previously in this file into the existing `breakpoint/breakpoint_utilities.py`

File: coral/breakpoint/breakpoint_types.py
Comment (suhas-r) [reply coral/breakpoint/breakpoint_types.py:26]: That makes a lot of sense, thanks for clarifying. I guess another option is rather than excluding certain lines, we could only parse lines beginning with `chr*` and skip all others? Assuming that's consistent regardless of the CN generator, it might be easier than trying to enumerate all other oddities?

Comment (jluebeck) [reply coral/breakpoint/breakpoint_types.py:36]: GRCh37 for instance just uses chromosomes named "1, 2, 3, ...". Another common use case could be samples with oncoviral integration (HPV, EBV, etc.) which also do not start with "chr". 

I suppose you could check if fields 1 and 2 are integers and use that to help decide if the line is valid non-header or not. In my opinion, only two file formats need to be supported - .cns and .bed. For anything else I think it is the user's responsibility to deal with :)

File: coral/breakpoint/breakpoint_graph.py
Comment (suhas-r) [reply coral/breakpoint/breakpoint_graph.py]: Method is unused now, will remove

Comment (suhas-r) [reply coral/breakpoint/breakpoint_graph.py]: I chatted with @kyzhu about this before commenting but I believe he mentioned that this functionality should be fully replaced by newer methods (namely `infer_discordant_edge_multiplicities`).

File: coral/breakpoint/breakpoint_graph.py
Comment (suhas-r) [reply coral/breakpoint/breakpoint_graph.py:146]: Added docstring for now. Agreed on the uninformative names, but I think these current names are used in a bunch of places and would rather update them all at once in a subsequent PR. 

(Personally, I would rather create a separate `Coordinate` namedtuple that encompasses `(chromosome, position, orientation)`, then stitch together multiple of these as necessary. But in any case, I think best for later)

File: coral/breakpoint/infer_breakpoint_graph.py
Comment (suhas-r) [reply coral/breakpoint/infer_breakpoint_graph.py]: renamed to `LongReadBamToBreakpointMetadata`, and added hyperparam docs

File: coral/cli.py
Comment (suhas-r) [reply coral/cli.py:1]: Hm, the various CLI commands + args already have help strings in their typer annotations, and the README gives additional documentation/examples for each mode. I'm not sure what addtl info there is to provide - do we just want to summarize the README again to reduce the chance of people missing it?

File: coral/breakpoint/breakpoint_graph.py
Comment (suhas-r) [reply coral/breakpoint/breakpoint_graph.py:900]: deprecated method, removing

File: coral/breakpoint/breakpoint_graph.py
Comment (suhas-r) [reply coral/breakpoint/breakpoint_graph.py:188]: I think I would like to clean up these edge definitions further and it would be more efficient to update their docstrings at that time - I've made an issue to track this: https://github.com/AmpliconSuite/CoRAL/issues/22.

File: coral/breakpoint/breakpoint_graph.py
Comment (suhas-r) [reply coral/breakpoint/breakpoint_graph.py:420]: Unused, removing. (Also if we do want to maintain this, I would probably need some additional info from @kyzhu to properly write this docstring as I'm not sure the short-read adjustments used here are well described in the paper vs. the long-read version I added.)

File: coral/breakpoint/breakpoint_graph.py
Comment (suhas-r) [reply coral/breakpoint/breakpoint_graph.py:28]: This is similar to #22 , I think this class would be far more concise + lend itself to simpler documentation with proper typing on what an `AmpliconInterval` is + associated modularization. Attempted to do so a while ago but ran into a host of typing errors due to inconsistent usage, so would prefer to punt for now

File: coral/breakpoint/breakpoint_types.py
Comment (suhas-r) [reply coral/breakpoint/breakpoint_types.py:36]: Reverting this to the previous check for now since no one's complained and doesn't seem like there's a perfect solution here

File: coral/breakpoint/breakpoint_graph.py
Comment (kyzhu) [reply coral/breakpoint/breakpoint_graph.py]: @suhas-r Please remove this function, for now, we only need to build breakpoint graph from long read. This function aims at building the graph from both short and long reads

File: coral/breakpoint/breakpoint_graph.py
Comment (kyzhu) [reply coral/breakpoint/breakpoint_graph.py]: @mattjones315 this function was already renamed - @suhas-r and I suggested ```enumerate_partitions``` or ```enumerate_clustering```

Comment (kyzhu) [reply coral/breakpoint/breakpoint_graph.py]: For your reference @suhas-r , basically, the goal of ```infer_discordant_edge_multiplicities``` is to infer the "multiplicity", i.e., the number of times a discordant edge can occur in **one ecDNA species**. However, due to the possibility that this discordant edge is shared by different ecDNA species, we actually infer an UPPER LIMIT on the number of times it can occur in any ecDNA species/rearranged genomes (as cycles or paths). Usually an overestimation would have minor impact as the same limit on other discordant edges may still constraint possible complex ecDNA structures, while an underestimation will disallow some true ecDNA structures. The estimation of multiplicities is purely based on read count supporting the discordant edges.

File: coral/breakpoint/breakpoint_graph.py
Comment (mattjones315) [reply coral/breakpoint/breakpoint_graph.py:28]: Sounds reasonable - can we make sure to record this feature request for a future PR?

File: coral/breakpoint/breakpoint_graph.py
Comment (mattjones315) [reply coral/breakpoint/breakpoint_graph.py:188]: Sounds great.

File: README.md
Comment (mattjones315) [inline README.md:1]: Great, thanks for updating the README. Let's make it _very_ clear that it only runs on hg38 right now. For the record, I think the next highest priority will be to enable usage for other builds.

File: coral/cycle_decomposition.py
Comment (mattjones315) [inline coral/cycle_decomposition.py:53]: All of these docstrings should have line length < 80 (put overflow on a new line with an indent)

File: coral/cycle_decomposition.py
Comment (mattjones315) [inline coral/cycle_decomposition.py:1]: nit: all the docstrings here go beyond 80 characters, let's get them within the limit. For an argument whose description goes beyond  80 characters, you can put the overflow on a new line with an indent.

**Review by mattjones315 — APPROVED** [review]
Really awesome work here Suhas!

Just two or three nit-picky comments to resolve before merging in. I'll approve here assuming you can resolve these before merging it in yourself. Congrats on your first major PR in CoRAL!

**Pull request merged**
